p2p/router.go

```diff
@@ -585,55 +587,83 @@ func (r *Router) openConnection(ctx context.Context, conn Connection) {
 func (r *Router) dialPeers() {
 	r.logger.Debug("starting dial routine")
 	ctx := r.stopCtx()
+
+	addresses := make(chan NodeAddress)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case address := <-addresses:
+					conn, err := r.dialPeer(ctx, address)
+					switch {
+					case errors.Is(err, context.Canceled):
+						return
+					case err != nil:
+						r.logger.Error("failed to dial peer", "peer", address, "err", err)
+						if err = r.peerManager.DialFailed(address); err != nil {
+							r.logger.Error("failed to report dial failure", "peer", address, "err", err)
+						}
+						return
+					}
+					defer conn.Close()
```

Comment:
Please note that this defer will only be run when the function exits, and not when the loop ends.
This could potentially leave connections open for a long time. If you'd like to keep this defer, make a closure for it.

Reply:
++ to function closure